### PR TITLE
Refactor model mapper: adding group-version-resource mapping

### PR DESCRIPTION
**Pull request by yue9944882**

### CODE REVIEW

File: e2e/src/test/groovy/io/kubernetes/client/e2e/util/ModelMapperTest.groovy [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+import io.kubernetes.client.openapi.models.V1Deployment
+import io.kubernetes.client.openapi.models.V1Pod
+import io.kubernetes.client.openapi.models.V1beta1CustomResourceDefinition
+import io.kubernetes.client.util.ClientBuilder
+import io.kubernetes.client.util.ModelMapper
+import spock.lang.Specification
+
+class ModelMapperTest extends Specification {
+
+    def "api-disocvery should work"() {
```
Comment (brendandburns) [inline e2e/src/test/groovy/io/kubernetes/client/e2e/util/ModelMapperTest.groovy:27]: `api-discovery should work` (typo)

File: util/src/main/java/io/kubernetes/client/util/ModelMapper.java [hunk 2]
```diff
@@ -42,16 +44,23 @@
...
+
   // Model's api-version midfix to kubernetes api-version
-  private static List<String> apiVersions = new ArrayList<>();
+  private static List<String> preBuiltApiVersions = new ArrayList<>();
+
+  private static Map<GroupVersionKind, Class<?>> classesByGVK = new ConcurrentHashMap<>();
+
+  private static Map<GroupVersionResource, Class<?>> classesByGVR = new ConcurrentHashMap<>();
+
+  private static Set<Discovery.APIResource> lastAPIDiscovery = new HashSet<>();
```
Comment (brendandburns) [inline util/src/main/java/io/kubernetes/client/util/ModelMapper.java:60]: Should this be a ConcurrentHashMap too? (you can use it like a `Set`)

File: util/src/main/java/io/kubernetes/client/util/ModelMapper.java [hunk 3]
```diff
@@ -114,33 +147,35 @@ public static void addModelMap(String group, String version, String kind, Class<
...
-   * Returns all the registered model classes.
+   * Gets the GVK by the given model class.
    *
-   * @return the all known classes
+   * @param clazz the clazz
+   * @return the group version kind by class
    */
-  public static Map<GroupVersionKind, Class<?>> getAllKnownClasses() {
-    return ImmutableMap.copyOf(classes);
+  public static GroupVersionKind getGroupVersionKindByClass(Class<?> clazz) {
```
Comment (brendandburns) [inline util/src/main/java/io/kubernetes/client/util/ModelMapper.java:163]: Do we expect this to be called a bunch? The performance of this is pretty slow, so we may want to make the map bi-directional.

**Comment by brendandburns** [comment]
Small  comments.

File: e2e/src/test/groovy/io/kubernetes/client/e2e/util/ModelMapperTest.groovy [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+        given:
+        def apiClient = ClientBuilder.defaultClient();
+        def discovery = new Discovery(apiClient)
+        ModelMapper.refresh(discovery);
+
+        expect:
+
```
[the remark this answers is not in the document]
Comment (yue9944882) [reply e2e/src/test/groovy/io/kubernetes/client/e2e/util/ModelMapperTest.groovy]: done

File: e2e/src/test/groovy/io/kubernetes/client/e2e/util/ModelMapperTest.groovy (hunk 1 above)
Comment (yue9944882) [reply e2e/src/test/groovy/io/kubernetes/client/e2e/util/ModelMapperTest.groovy]: fixed

File: util/src/main/java/io/kubernetes/client/util/ModelMapper.java (hunk 2 above)
Comment (yue9944882) [reply util/src/main/java/io/kubernetes/client/util/ModelMapper.java:62]: the map is actually thread-safe b/c it's immutable. we're replace the `lastAPIDiscovery` w/ a new immutable map upon every refreshing

File: util/src/main/java/io/kubernetes/client/util/ModelMapper.java (hunk 3 above)
Comment (yue9944882) [reply util/src/main/java/io/kubernetes/client/util/ModelMapper.java:165]: yeah ideally we should make this bi-directional. i was making it a GVK->class map b/c class->GVK doesnt look like a frequent case i suppose.. i left a TODO comment here and will do this in a follow-up. wdyt?

**Pull request merged**